backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala

```diff
@@ -954,6 +954,43 @@ trait StandardAsyncExecutionActor
     }
   }
 
+  /**
+   * Merge key-value pairs from previous job execution attempt with incoming pairs from backend and store them in
+   * the database. In case when there are key-value pairs with the same key name among those from previous attempt and
+   * coming from backend, backend values have higher precedence.
+   * `FailedRetryCountKey` is handled separately outside of this method 
+   *
+   */
+  private def saveKvPairsForNextAttempt(backendExecutionStatus: Future[ExecutionHandle]): Future[ExecutionHandle] = {
+    val kvsFromPreviousAttempt = jobDescriptor.prefetchedKvStoreEntries
+      // We exclude FailedRetryCountKey here since it will be incremented and saved separately, after we ensure
+      // that the next retry will actually occur
+      .filter {
+        case (key, value) => key != BackendLifecycleActorFactory.FailedRetryCountKey && value.isInstanceOf[KvPair]
+      }
+      .map {
+        case (key, value) =>
+          val futureKvJobKey = KvJobKey(jobDescriptor.key.call.fullyQualifiedName, jobDescriptor.key.index, jobDescriptor.key.attempt + 1)
+          val kvPairVal = value.asInstanceOf[KvPair]
+          key -> kvPairVal.copy(key = kvPairVal.key.copy(jobKey = futureKvJobKey))
+      }
+
+    backendExecutionStatus flatMap {
+      case failed: FailedExecutionHandle =>
+        failed.kvPairsToSave match {
+          case Some(kvPairs) =>
+            val kvsFromBackend = kvPairs.map(kvPair => kvPair.key.key -> kvPair).toMap
+            val mergedKvs = (kvsFromPreviousAttempt ++ kvsFromBackend) map {
+              case (_, value) => value
+            }
```

Comment:
I think this is equivalent
```
            val mergedKvs = (kvsFromPreviousAttempt.values ++ kvPairs)
```

Reply:
@mcovarr looks like this would simply append one sequence to another, but I wanted to enforce key uniqueness and precedence of values, which came from backend, over values from previous attempt, hence joining two `Map[String, KvPair]`

Reply:
Ah right, as it says in the scaladoc. 🙂